### DISCUSSION TITLE
flush() not working as expected

### DISCUSSION THREAD

**Discussion by kaushal02** [gabime/spdlog#2982] [Q&A] [answered]
I have a `FileLogger` class as follows:

```
class FileLogger {
 public:
  FileLogger(const std::string &service_name, int max_files,
             int max_file_size, const char *log_file);
  void Log(level::LogLevel l, const std::string &message) const override;
  void Flush() const override;
  static void FlushSink();

 private:
  static std::mutex sink_mutex;
  static std::optional<spdlog::sink_ptr> sink;
  std::shared_ptr<spdlog::async_logger> logger;
};
```

the definitions are as follows:
```
FileLogger::FileLogger(const std::string &service_name,
                       int max_files, int max_file_size, const char *log_file) {
  if (!sink.has_value()) {
    std::lock_guard<std::mutex> lock(sink_mutex);
    if (!sink.has_value()) {
      sink = std::make_shared<spdlog::sinks::rotating_file_sink_mt>(
          log_file, max_file_size, max_files);
    }
  }

  auto tp = spdlog::thread_pool();
  if (tp == nullptr) {
    spdlog::init_thread_pool(spdlog::details::default_async_q_size, 1U);
    tp = spdlog::thread_pool();
  }
  logger = std::make_shared<spdlog::async_logger>(
      service_name, sink.value(), tp,
      spdlog::async_overflow_policy::overrun_oldest);
  spdlog::initialize_logger(logger);
  logger->set_pattern("[%Y-%m-%d %T.%e %z] [%^%l%$] [%n] %v");
}

void FileLogger::Log(level::LogLevel l, const std::string &message) const {
  logger->log(level::GetSpdLogLevel(l), message);
}

void FileLogger::Flush() const {
  logger->flush();
}

void FileLogger::FlushSink() {
  if (sink.has_value()) {
    sink.value()->flush();
  }
}
```

I am trying to write unittests for this class, but the logs do not get flushed even after I use `FileLogger::Flush()` and `FileLogger::FlushSink()`. What am I doing wrong?

```
  auto logger1 = CreateLogger(config);
  auto logger2 = CreateLogger(config);
  logger2->Log(LogLevel::kError, "error message is logged");
  logger1->Log(LogLevel::kInfo, "info message is logged");
  logger2->Log(LogLevel::kInfo, "info message is logged");

  logger1->Flush();
  logger2->Flush();
  FileLogger::FlushSink();
```

Even at this point, the logs do not appear in the logfile, they appear only after the application exits. Is there something else I can do to force flush the logs? Thanks!

**Comment by tt4g** [comment, marked as answer]
The async logger is also asynchronous for flush operations.
The caller does not receive notification of flush completion.

See also: #1696

EDIT: `FileLogger::FlushSink()` is a synchronous operation, but if it is called before the log message reaches sink, there is no log to flush.

**Reply by kaushal02** [reply]
Okay so I have multiple async_logger instances as you can se in the above unittests. It's working fine if I do the below:
```
  auto logger1 = CreateLogger(config);
  auto logger2 = CreateLogger(config);
  logger2->Log(LogLevel::kError, "error message is logged");
  logger1->Log(LogLevel::kInfo, "info message is logged");
  logger2->Log(LogLevel::kInfo, "info message is logged");

  spdlog::details::registry::instance().flush_all();
  spdlog::shutdown();
```

Wondering if there is some API to undo the effects of `shutdown()` so that I can create new file_loggers even after the `shutdown()` is called.

**Reply by tt4g** [reply]
After removing all loggers by `spdlog::drop_all()`, remove the old thread-pool by `spdlog::init_thread_pool()`.
The thread-pool completes all remaining tasks in the destructor.

`spdlog::drop_all()` differs from `spdlog::shutdown()` in that it releases only loggers registered in the spdlog registry.

Note `spdlog::drop_all()` delete `spdlog::default_logger()`.
